### PR TITLE
[cron testing] Diff nodejs tests and send email

### PR DESCRIPTION
**Pull request by chejennifer**
Update cron testing job to diff nodejs test results against a golden set and send email if diffs are found.

Changes:
- add new send_email tool that takes a recipient and email content and sends the email to the recipient
- update nodejs_query_differ tool to diff against a golden set instead of previous run (will have a separate PR for updating goldens)
- update website cron testing job to run nodejs_query_differ after generating nodejs query results and send email if diffs are found

Email looks like: 

Screenshot 2024-04-15 at 12 10 11 PM

### CODE REVIEW

File: tools/nl/nodejs_query_differ/differ.py [hunk 1]
```diff
@@ -21,6 +22,12 @@
 
 _GCS_BUCKET = 'datcom-website-periodic-testing'
 _OUTPUT_FILE = 'differ_results.json'
+_GOLDEN_FOLDER = 'golden'
+_DIFF_SUCCESS_MSG = 'Success'
+_EMAIL_SUBJECT_KEY = 'subject'
+_EMAIL_SUBJECT_TEMPLATE = 'Nodejs Query Test Failure in {env}'
+_EMAIL_MESSAGE_KEY = 'message'
+_EMAIL_MESSAGE_TEMPLATE = 'There were diffs found when testing Nodejs Query results against goldens in {env}.<br><br><b>Goldens folder</b>: {goldens_path}<br><b>Nodejs Query Results</b>: {test_path}<br><b>Diff results</b>: {results_path}'
```
Comment (pradh) [inline tools/nl/nodejs_query_differ/differ.py:30]: Nice, could you add an example diff email screenshot?

File: tools/nl/nodejs_query_differ/differ.py [hunk 2]
```diff
@@ -21,6 +22,12 @@
 
 _GCS_BUCKET = 'datcom-website-periodic-testing'
 _OUTPUT_FILE = 'differ_results.json'
+_GOLDEN_FOLDER = 'golden'
+_DIFF_SUCCESS_MSG = 'Success'
+_EMAIL_SUBJECT_KEY = 'subject'
+_EMAIL_SUBJECT_TEMPLATE = 'Nodejs Query Test Failure in {env}'
```
Comment (pradh) [inline tools/nl/nodejs_query_differ/differ.py:28]: super nit:   maybe --  `[{env}] Failure: Nodejs Query Test`

File: tools/nl/nodejs_query_differ/differ.py [hunk 3]
```diff
@@ -30,82 +37,120 @@
...
+    logging.info(f'Diff results saved to gcs path: {results_path}')
+    return results_path
+  else:
+    with open(_OUTPUT_FILE, 'w') as f:
+      f.write(json.dumps(results))
+    logging.info(f'Diff results saved locally to: {_OUTPUT_FILE}')
+    return _OUTPUT_FILE
+
+
+def output_failure_email(results, golden_folder, test_folder, results_path):
```
Comment (pradh) [inline tools/nl/nodejs_query_differ/differ.py:116]: Do you have an example failure diff result?

**Review by pradh — APPROVED** [review]
Very nice!

File: tools/nl/nodejs_query_differ/differ.py (hunk 3 above)
Comment (chejennifer) [reply tools/nl/nodejs_query_differ/differ.py:116]: I have a screenshot in the PR description?

File: tools/nl/nodejs_query_differ/differ.py (hunk 1 above)
Comment (chejennifer) [reply tools/nl/nodejs_query_differ/differ.py:30]: I have an example in the PR description, where did you want this example?

Comment (pradh) [reply tools/nl/nodejs_query_differ/differ.py:30]: Sorry, missed it.  Should we have the paths be like:

https://console.cloud.google.com/storage/browser/datcom-website-periodic-testing/dev

So they are clickable in the browser?

File: tools/nl/nodejs_query_differ/differ.py (hunk 3 above)
Comment (pradh) [reply tools/nl/nodejs_query_differ/differ.py:116]: I was looking for the file contents here.

Found it from your screenshot image, after scrolling for a bit:

`{"bar": "Success", "disaster": "Success", "map_rank": "Success", "scatter": "Success", "scatter_non_bard": "Success", "timeline": "Success", "timeline_all": "Success"}`

Can we do indent=2?

Also, how does a failure look?

**Review by pradh — APPROVED** [review]
Thanks.

File: tools/nl/nodejs_query_differ/differ.py (hunk 1 above)
Comment (chejennifer) [reply tools/nl/nodejs_query_differ/differ.py:31]: done

File: tools/nl/nodejs_query_differ/differ.py (hunk 3 above)
Comment (pradh) [reply tools/nl/nodejs_query_differ/differ.py:118]: Hmm, ok - what would the next step to debug be - regenerate goldens and review?

Comment (pradh) [reply tools/nl/nodejs_query_differ/differ.py:118]: Might be helpful to have a paragraph on "steps to take on diff failure" in go/dcrelease or somewhere easily accessible.

**Comment by pradh** [comment]
Thanks for all the edits!  This is super helpful / valuable!

**Review by pradh — APPROVED** [review]

**Comment by chejennifer** [comment]
Thanks for the review, I'll add links to the playbook entry once that's in (https://critique.corp.google.com/cl/625759580)

**Pull request merged**